src/core/installAddon.js

```diff
@@ -228,24 +233,42 @@ const userAgentOSToPlatform = {
  *
  * Parameter types:
  *
- * import type { AddonType } from 'src/core/types/addons';
- * import type { UserAgentInfoType } from 'src/core/reducers/api';
+ * import type { AddonType } from 'core/types/addons';
+ * import type { ReactRouterLocation } from 'core/types/router';
+ * import type { UserAgentInfoType } from 'core/reducers/api';
  *
  * type FindInstallUrlParams = {|
+ *   appendSource?: boolean,
+ *   defaultSource?: string,
+ *   location?: ReactRouterLocation,
  *   platformFiles: $PropertyType<AddonType, 'platformFiles'>,
- *   src?: string,
  *   userAgentInfo: UserAgentInfoType,
  * |};
  *
  */
-export const findInstallURL = ({ platformFiles, userAgentInfo, src }) => {
+export const findInstallURL = ({
+  appendSource = true,
+  defaultSource,
+  location,
+  platformFiles,
+  userAgentInfo,
+}) => {
   if (!platformFiles) {
     throw new Error('The platformFiles parameter is required');
   }
   if (!userAgentInfo) {
     throw new Error('The userAgentInfo parameter is required');
   }
 
+  let source;
+  if (appendSource) {
+    if (!location) {
+      throw new Error(
+        'The location parameter is required when appendSource is true');
+    }
+    source = location.query.src || defaultSource;
```

Comment:
I get what this is doing but the logic might feel a bit clearer to me if it were restructured. As you're already using a `let` and redefining `source` (which I think is fine here), maybe:

```js
let source = defaultSource;
if (appendSource) {
  if (!location) {
    throw new Error(
      'The location parameter is required when appendSource is true');
  }
  source = location.query.src;
}
```

I think that's just a bit easier to follow.

Reply:
Oh wait, I see... if `appendSource` is `false` we shouldn't be setting it at all. Should that ever really happen though? If so... I guess the original code is cool but I just had to read it a few times to totally follow it.